yew/src/dev/mod.rs

```diff
@@ -0,0 +1,228 @@
+//! Developer tools.
+//! These communicate with a debugging server (normally a browser extension) over a WebSocket connection.
+
+use serde::Serialize;
+use serde_json;
+
+#[cfg(test)]
+pub use messages::tests as message_tests;
+
+#[cfg(feature = "web_sys")]
+use web_sys;
+
+#[cfg(feature = "std_web")]
+use stdweb;
+
+pub mod messages;
+
+/// Stores a connection to the DevTools server.
+#[derive(Debug, Clone)]
+pub struct DebuggerConnection {
+    #[cfg(feature = "web_sys")]
+    /// Public only for testing.
+    pub ws: web_sys::WebSocket,
+    #[cfg(feature = "std_web")]
+    /// Public only for testing.
+    pub ws: stdweb::web::WebSocket,
+    message_queue: Vec<String>,
+}
+
+/// Describes the state of a WebSocket connection.
+#[derive(Debug)]
+pub enum ConnectionState {
+    /// The socket has successfully connected.
+    Connected,
+    /// The socket has not successfully connnected.
+    CouldntConnect,
+}
+
+impl std::future::Future for DebuggerConnection {
+    type Output = Self;
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        match self.ws.ready_state() {
+            0 => std::task::Poll::Pending,
+            1 => std::task::Poll::Ready(self.clone()),
+            2 => std::task::Poll::Pending,
+            3 => std::task::Poll::Ready(self.clone()),
+            _ => std::task::Poll::Pending,
+        }
+    }
+}
+
+/// A debugger is capable of sending messages over a WebSocket connection.
+pub trait DebuggerMessageQueue<T>
+where
+    T: Serialize,
+{
+    /// Adds a message to the message queue.
+    fn queue_message(&mut self, message: T);
+}
+
+/// Sends messages.
+pub trait DebuggerMessageSend {
+    /// Send all the messages in the queue.
+    fn send_messages(&mut self);
+}
+
+impl<T: Serialize> DebuggerMessageQueue<T> for DebuggerConnection {
+    fn queue_message(&mut self, message: T) {
+        self.message_queue
+            .push(serde_json::to_string(&message).unwrap());
+    }
+}
+
+impl DebuggerMessageSend for DebuggerConnection {
+    fn send_messages(&mut self) {
```

Comment:
When is this called? I would think it would be called after adding messages to the queue